### ISSUE TITLE
Fencing the command buffers

### ISSUE THREAD

**Issue by kvark** [gfx-rs/gfx#688]
Suggested by @tomaka :

> what I think should be done for vulkan is that you create a fence each time you submit a command buffer
> this fence is returned in a `Fence` object that holds smart pointers to the resources used by the command buffer
> and Fence's destructor waits for the fence, or you can use it when submitting another command buffer behind the first one
> `submit_command_buffer(after: Option<Fence>, commands: &CommandsBuffer)`
> in which case it doesn't wait
>  fact it would be `submit_command_buffer(after: Option<Fence>, commands: &CommandsBuffer) Fence`
> and the smart pointers in the Fence that you pass as parameter are moved to the Fence that is returned
> at the end of the frame, you wait on the fence
> and it destroys the resources

**Comment by csherratt** [comment]
I have been working on a library that could be useful to build a `Fence` with.

https://github.com/csherratt/pulse

**Comment by kvark** [comment]
This is pretty cool!

**Comment by csherratt** [comment]
I have been thinking about this some more, mostly because I need some support for fencing of commands to reduce the latency for `gfx_vr`. The recommended method suggested by Carmack is to draw a small triangle after the frame swap and wait on a fence immediately after it. This stops the GPU from being to aggressive with driver side queueing.

I think overall the approach of adding an optional fence is a good approach. This allows `gfx` to intelligently use `glWaitSync​`, and if the user wants to stall the CPU (to say avoid accessing a memory mapped buffer, or to reduce latency in VR) they can.

The only issue I see with cpu stalling is that the `glClientWaitSync` is not a very nice library for dealing with multi-threaded work. We can't move the OpenGL context around as needed, so we will somehow need to manager multiple threads waiting on multiple fences. Which sounds incredibility messy.

Maybe v1 can just do the fencing and possibly a `glClientWaitSync` using the `device` directly. That solves the multi-threaded issue by making it impossible (for now).

**Comment by kvark** [comment]
This is actually required for the Vulkan backend #629 . A fence per command buffer not only signifies when the used resources are no longer needed, but it also tells us when that command buffer can be re-used.

**Comment by kvark** [comment]
Not happened in pre-ll
